service/client/cadence/client.go

```diff
@@ -227,9 +240,24 @@ func (t *cadenceClient) ListWorkflow(
 func (t *cadenceClient) QueryWorkflow(
 	ctx context.Context, valuePtr interface{}, workflowID string, runID string, queryType string, args ...interface{},
 ) error {
-	qres, err := queryWorkflowWithStrongConsistency(t, ctx, workflowID, runID, queryType, args)
-	if err != nil {
-		return err
+	var qres encoded.Value
+	var err error
+
+	attempt := 1
+	for attempt <= t.queryWorkflowFailedRetryPolicy.MaximumAttempts {
+		qres, err = queryWorkflowWithStrongConsistency(t, ctx, workflowID, runID, queryType, args)
+		if err != nil {
+			if t.isQueryFailedError(err) {
+				if attempt == t.queryWorkflowFailedRetryPolicy.MaximumAttempts {
```

Comment:
Nit: I don't think we need another check here, since it'a already checked by the for loop condition?

Although we will have to check `err != nil` again outside of the loop

Reply:
Boils down to the same thing, but it might be a little cleaner. Done